pkg/providers/vault/provider.go

```diff
@@ -0,0 +1,376 @@
+	roleName := attrib["roleName"]
+	if roleName == "" {
+		return errors.Errorf("missing vault role")
+	}
+	p.VaultRole = roleName
+
+	glog.V(2).Infof("vault: roleName %s", p.VaultRole)
+
+	p.VaultAddress = attrib["vaultAddress"]
+	if p.VaultAddress == "" {
+		p.VaultAddress = defaultVaultAddress
+	}
+	glog.V(2).Infof("vault: vault address %s", p.VaultAddress)
+
+	// One of the following variables should be set when vaultSkipTLSVerify is false.
+	// Otherwise, system certificates are used to make requests to vault.
+	p.VaultCAPem = attrib["vaultCAPem"]
+	p.VaultCACert = attrib["vaultCACertPath"]
+	p.VaultCAPath = attrib["vaultCADirectory"]
+	// Vault tls server name.
+	p.VaultServerName = attrib["vaultTLSServerName"]
+
+	if s := attrib["vaultSkipTLSVerify"]; s != "" {
+		b, err := strconv.ParseBool(s)
+		if err != nil {
+			return err
+		}
+		p.VaultSkipVerify = b
+	}
+
+	p.VaultK8SMountPath = attrib["vaultKubernetesMountPath"]
+	if p.VaultK8SMountPath == "" {
+		p.VaultK8SMountPath = defaultVaultKubernetesMountPath
+	}
+
+	p.KubernetesServiceAccountPath = attrib["vaultKubernetesServiceAccountPath"]
+	if p.KubernetesServiceAccountPath == "" {
+		p.KubernetesServiceAccountPath = defaultKubernetesServiceAccountPath
+	}
+
+	keyValueObjects := []KeyValueObject{}
+	objectsStrings := attrib["objects"]
+	fmt.Printf("objectsStrings: [%s]\n", objectsStrings)
+
+	var objects StringArray
+	err = yaml.Unmarshal([]byte(objectsStrings), &objects)
+	if err != nil {
+		fmt.Printf("unmarshall failed for objects")
+		return err
+	}
+	fmt.Printf("objects: [%v]", objects.Array)
+	for _, object := range objects.Array {
+		fmt.Printf("unmarshal object: [%s]\n", object)
+		var keyValueObject KeyValueObject
+		err = yaml.Unmarshal([]byte(object), &keyValueObject)
+		if err != nil {
+			fmt.Printf("unmarshall failed for keyValueObjects at index")
+			return err
+		}
+
+		keyValueObjects = append(keyValueObjects, keyValueObject)
+	}
+
+	for _, keyValueObject := range keyValueObjects {
+		content, err := p.GetKeyValueObjectContent(ctx, keyValueObject.ObjectPath, keyValueObject.ObjectName, keyValueObject.ObjectVersion)
+		if err != nil {
+			return err
+		}
+		objectContent := []byte(content)
+		if err := ioutil.WriteFile(path.Join(targetPath, keyValueObject.ObjectPath), objectContent, permission); err != nil {
+			return errors.Wrapf(err, "secrets-store csi driver failed to write %s at %s", keyValueObject.ObjectPath, targetPath)
+		}
+		glog.V(0).Infof("secrets-store csi driver wrote %s at %s", keyValueObject.ObjectPath, targetPath)
+	}
+
+	return nil
+}
+
+// GetKeyValueObjectContent get content of the vault object
+func (p *Provider) GetKeyValueObjectContent(ctx context.Context, objectPath string, objectName string, objectVersion string) (content string, err error) {
+	// Read the jwt token from disk
+	jwt, err := readJWTToken(p.KubernetesServiceAccountPath)
```

Comment:
Is this using the service account of the pod running the _secrets-store-csi_ program on each node (via a _DaemonSet_)? Ideally we'd use the service account token of the pod requesting the mounting of the secret content, so that we can bind our Vault roles to those service accounts, and not to the service account with which the _secrets-store-csi_ pod runs.

It looks like we wind up here when `(*secretstore.nodeServer).NodePublishVolume` gets invoked. That sounds like it's running in the daemon pod, and not in the context of the mounting pod.

If @anubhavmishra isn't sure, @ritazh may know.

Reply:
Yea, that is correct. Currently, this is using the service account JWT token of the daemon set running the driver. It should be running in the context of the mounting pod. I wasn't quite sure how to best implement it at the time.

Reply:
Thank you for confirming. Do you have an implementation plan now, or are you open to suggestions? I have an idea, but it comes from ignorance of what the CSI allows these drivers to do.

Should we elevate this to its own issue?